_drafts/cortexm-and-clang.md

```diff
@@ -0,0 +1,971 @@
+image:
+tag: [cortex-m]
+author: chris
+---
+
+If you have ever done any software development that involves Apple products, FreeBSD, WebAssembly, or Rust you have made use of a wildly popular compiler toolchain known as LLVM. However, in the embedded world, LLVM and the open source C-language compiler built on top of it (Clang) still do not get a lot of attention.
+
+<!-- excerpt start -->
+
+In this article, I hope I can convince you that adding a Clang build target to your project can be
+a relatively simple and useful endeavor. We will cover how to take advantage of some of the
+rich features shipped with the toolchain and identify some memory leaks and mutex deadlocks in an
+example project at compile time! We will also walk through a step-by-step example of updating a GCC
+based ARM Cortex-M project to compile you need to take to cross-compile with LLVM/Clang.
+
+<!-- excerpt end -->
+
+> Note: While the focus of the article will be using LLVM/Clang with ARM Cortex-M embedded devices, the
+> general overview of LLVM as well as the features leveraged can be applied to any architecture
+> supported by the toolchain.
+
+{:.no_toc}
+
+## Table of Contents
+
+<!-- prettier-ignore -->
+* auto-gen TOC:
+{:toc}
+
+## Background
+
+The advent of open source compilers and toolchains traces back over 30 years to 1987 and the
+release of GCC 1.0. The origins of the toolchain were for smooth compilation of GNU open source
+software written in C. The toolchain grew organically over time to target numerous architectures and support a
+variety of different languages. As a consequence, many aspects of the codebase have became complex to extend and work with.
+
+The LLVM project is the "new" open source compiler on the block. It began as a research project at the
+University of Illinois at Urbana–Champaign in 2000[^12]. The early goals of the project were to create a
+more modular compiler toolchain by designing a collection of standalone libraries. The LLVM project really took root in 2005 when Apple hired one of the original designers, Chris
+Lattner, and invested heavily in making LLVM the default used for its own developer ecosystem. This
+bore Clang (a C, C++, & Objective-C) frontend compiler to LLVM and modern day GNU GCC competitor.
+
+Today, the LLVM toolchain has become ubiquitous in the software world and in recent years it has seen
+adoption in the embedded world as well. The favorable license in comparison to the GNU
+toolchain (Apache 2 vs GPL3) along with the more modular architecture has made the LLVM toolchain a
+default choice in the proprietary embedded compiler market. ARM scrapped their original compiler in
+favor of an LLVM/Clang based one with the release of ARM Compiler 6 (`armclang`) and more recently
+SEGGER built an ARM compiler based on the LLVM/Clang toolchain as
+well[^11].
+
+It's clear LLVM/Clang is here to stay and as we will discover below there are already quite a few
```

Comment:
```suggestion
It's clear LLVM/Clang is here to stay and, as we will discover below, there are already quite a few
```

Regardless, I don't love this sentence. 

It's clear LLVM/Clang is here to stay. As such, there are already quite a few ways it can be utilized for embedded development today, which we will discover below!

Reply:
I removed it. I think the replacement section you suggested captures what I was trying to achieve